core/src/letter_distribution.c

```diff
@@ -153,47 +153,51 @@ uint8_t human_readable_letter_to_machine_letter(
 // Convert a string of arbitrary characters into an array of machine letters,
 // returning the number of machine letters. This function does not allocate
 // the ml array; it is the caller's responsibility to make this array big
-// enough.
+// enough. This function will return -1 if it encounters an invalid letter.
 // Note: This is a slow function that should not be used in any hot loops.
 int str_to_machine_letters(const LetterDistribution *letter_distribution,
                            const char *str, bool allow_played_through_marker,
                            uint8_t *mls) {
 
   int num_mls = 0;
   int num_bytes = string_length(str);
-  int i = 0;
-  int prev_i = -1;
-  while (i < num_bytes) {
-    for (int j = i + letter_distribution->max_tile_length; j > i; j--) {
-      if (j > num_bytes) {
-        continue;
+  // Use +1 for the null terminator
+  char current_letter[MAX_LETTER_CHAR_LENGTH + 1];
+  int current_letter_byte_index = 0;
+  bool building_multichar_letter = false;
+
+  for (int i = 0; i < num_bytes; i++) {
+    char current_char = str[i];
+    if (current_char == '[') {
+      building_multichar_letter = true;
+    } else if (current_char == ']') {
+      building_multichar_letter = false;
+    } else {
+      if (current_letter_byte_index == MAX_LETTER_CHAR_LENGTH) {
+        // multichar exceeded max char length
+        return -1;
       }
-      // possible letter goes from index i to j. Search for it.
-      char possible_letter[MAX_LETTER_CHAR_LENGTH];
-      memory_copy(possible_letter, str + i, j - i);
-      possible_letter[j - i] = '\0';
+      current_letter[current_letter_byte_index] = current_char;
+      current_letter_byte_index++;
+      current_letter[current_letter_byte_index] = '\0';
```

Comment:
some cases i'm not seeing handled:
- `[` while already `building_multichar_letter` (not an error?)
- reaching end of `num_bytes` while still `building_multichar_letter` (not an error?)
- starting with `[]` or `]` (undefined behavior/crashes because `current_letter` is uninitialized?)
- `[]` or `]` after first `ml` while currently `!building_multichar_letter` (repeats the last single- or multichar letter?)
- out-of-bounds for `ml` because size is unknown? (already noted in comment)

writing this `\0` only after the whole series, rather than after each byte, may fix some of the above, but a different approach is to make `human_readable_letter_to_machine_letter` take a pointer and length and use something like `memcmp` (or `strncmp`?) and skip copying `current_letter` at all

Reply:
Sorry, I'm a bit confused. What does out-of-bounds for `ml` mean? What does size refer to? Which comment are you referring to, your own in this review or a comment in the code?

Reply:
this comment:
https://github.com/jvc56/MAGPIE/blob/b9445b4e166ec3b50cbc2a262289b20660b829f5/core/src/letter_distribution.c#L153-L160

> it is the caller's responsibility to make this array big enough.

for example,
https://github.com/jvc56/MAGPIE/blob/b9445b4e166ec3b50cbc2a262289b20660b829f5/core/src/rack.c#L84-L98

with a long `rack_string` would overwrite `mls[MAX_BAG_SIZE]` (likely corrupting stack) before getting to `if (num_mls > MAX_BAG_SIZE) { log_fatal(...); }`, because `str_to_machine_letters` doesn't know the size it's allowed to write.

an unintuitive way to secure `set_rack_to_string` notes that _currently_ one output byte corresponds to one or more input chars, and `str_to_machine_letters` will not write to `mls[string_length(rack_string)]`:
```c
   uint8_t mls[MAX_BAG_SIZE];
   size_t len = string_length(rack_string);
   if (len > MAX_BAG_SIZE) {
     log_fatal("rack overflow: %zu\n", len); /* is %zu gcc-specific? */
   }
   int num_mls =
       str_to_machine_letters(letter_distribution, rack_string, false, mls);
```

Reply:
Ah, I see, I will work on these changes.